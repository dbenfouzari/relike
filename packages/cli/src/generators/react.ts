import { paramCase, snakeCase } from "change-case";
import fs from "fs";
import path from "path";

import { replace, replaceFileName } from "../utils/template";

/**
 * Generates component index file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateIndexFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateIndexFile = (key: string, componentName: string) => {
  const templateIndexPath = path.resolve(__dirname, "../templates/index/index.ts.txt");

  const templateIndex = fs.readFileSync(templateIndexPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateIndexPath, key, componentName)),
    content: replace(templateIndex, key, componentName, paramCase),
  };
};

/**
 * Generates component style file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateStyleFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateStyleFile = (key: string, componentName: string) => {
  const templateStylePath = path.resolve(__dirname, "../templates/react-style/__name__.module.scss.txt");

  const templateStyle = fs.readFileSync(templateStylePath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateStylePath, key, componentName)),
    content: replace(templateStyle, key, componentName, snakeCase),
  };
};

/**
 * Generates component story file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateStoryFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateStoryFile = (key: string, componentName: string) => {
  const templateStoryPath = path.resolve(__dirname, "../templates/react-story/__name__.stories.tsx.txt");

  const templateStory = fs.readFileSync(templateStoryPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateStoryPath, key, componentName)),
    content: replace(templateStory, key, componentName),
  };
};

/**
 * Generates component test file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateTestFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateTestFile = (key: string, componentName: string) => {
  const templateTestPath = path.resolve(__dirname, "../templates/react-test/__name__.test.tsx.txt");

  const templateTest = fs.readFileSync(templateTestPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateTestPath, key, componentName)),
    content: replace(templateTest, key, componentName),
  };
};

/**
 * Generates component file
 *
 * @param   {string}  key           The key to replace
 * @param   {string}  componentName The component name
 * @param   {boolean} hasStyle      Does the component have style
 * @example
 * generateComponentFile("name", name, flags.style);
 * @returns {any}                   File path and content.
 */
export const generateComponentFile = (key: string, componentName: string, hasStyle: boolean) => {
  const templateComponentPath = hasStyle
    ? path.resolve(__dirname, "../templates/react-component/__name__(if:style).tsx.txt")
    : path.resolve(__dirname, "../templates/react-component/__name__.tsx.txt");

  const templateComponent = fs.readFileSync(templateComponentPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateComponentPath, key, componentName)),
    content: replace(templateComponent, key, componentName),
  };
};
