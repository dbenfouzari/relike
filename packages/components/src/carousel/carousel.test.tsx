import "intersection-observer";

import { fireEvent, render, screen } from "@testing-library/react";
import { CSSProperties } from "react";

import Colors from "../colors";
import Carousel from "./carousel";

const styles: Record<string, CSSProperties> = {
  element: {
    width: "20rem",
    height: "12rem",
    borderRadius: 4,
    backgroundColor: Colors.grey[200].toRGBA(),
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
  },
};

const mockScrollLeft = jest.fn();
const mockScrollRight = jest.fn();

jest.mock("./useCarousel", () =>
  jest.fn(() => ({
    leftIndicator: true,
    rightIndicator: true,
    /**
     * Component ref creator.
     *
     * @returns {any} The ref
     * @example
     * <div ref={itemRef} />
     */
    itemRef: () => jest.fn(),
    scrollLeft: mockScrollLeft,
    scrollRight: mockScrollRight,
  })),
);

/**
 * Set up tests
 *
 * @returns {any} The result
 * @example
 * setup()
 */
const setup = () => {
  return render(
    <Carousel>
      <div key={1} style={styles.element}>
        Example 1
      </div>
      <div key={2} style={styles.element}>
        Example 2
      </div>
      <div key={3} style={styles.element}>
        Example 3
      </div>
      <div key={4} style={styles.element}>
        Example 4
      </div>
      <div key={5} style={styles.element}>
        Example 5
      </div>
      <div key={6} style={styles.element}>
        Example 6
      </div>
      <div key={7} style={styles.element}>
        Example 7
      </div>
    </Carousel>,
  );
};

describe("Carousel", function () {
  it("should render successfully", () => {
    setup();

    expect(screen.getByTestId("carousel")).toMatchSnapshot();
  });

  it("should handle next click", () => {
    setup();

    const rightIndicator = screen.getByTestId("right");
    fireEvent.click(rightIndicator);

    expect(mockScrollRight).toHaveBeenCalled();
  });

  it("should handle prev click", () => {
    setup();

    const leftIndicator = screen.getByTestId("left");
    fireEvent.click(leftIndicator);

    expect(mockScrollLeft).toHaveBeenCalled();
  });
});
