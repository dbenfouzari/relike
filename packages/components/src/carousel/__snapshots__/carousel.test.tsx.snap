// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel should render successfully 1`] = `
<div
  class="wrapper"
  data-testid="carousel"
>
  <button
    class="wrapper wrapper-0-2-1 wrapper-d0-0-2-2 wrapper__clickable button"
    data-testid="left"
    tabindex="0"
  >
    <span
      class="jsx-2787895825 icon material-icons"
    >
      chevron_left
    </span>
  </button>
  <ul
    class="inner"
    role="tablist"
    style="gap: 12px;"
    tabindex="0"
  >
    <li
      aria-label="Slide 1 of 7"
      class="element"
      role="tab"
    >
      <div
        style="width: 20rem; height: 12rem; border-radius: 4px; background-color: rgb(238, 238, 238); display: flex; align-items: center; justify-content: center;"
      >
        Example 1
      </div>
    </li>
    <li
      aria-label="Slide 2 of 7"
      class="element"
      role="tab"
    >
      <div
        style="width: 20rem; height: 12rem; border-radius: 4px; background-color: rgb(238, 238, 238); display: flex; align-items: center; justify-content: center;"
      >
        Example 2
      </div>
    </li>
    <li
      aria-label="Slide 3 of 7"
      class="element"
      role="tab"
    >
      <div
        style="width: 20rem; height: 12rem; border-radius: 4px; background-color: rgb(238, 238, 238); display: flex; align-items: center; justify-content: center;"
      >
        Example 3
      </div>
    </li>
    <li
      aria-label="Slide 4 of 7"
      class="element"
      role="tab"
    >
      <div
        style="width: 20rem; height: 12rem; border-radius: 4px; background-color: rgb(238, 238, 238); display: flex; align-items: center; justify-content: center;"
      >
        Example 4
      </div>
    </li>
    <li
      aria-label="Slide 5 of 7"
      class="element"
      role="tab"
    >
      <div
        style="width: 20rem; height: 12rem; border-radius: 4px; background-color: rgb(238, 238, 238); display: flex; align-items: center; justify-content: center;"
      >
        Example 5
      </div>
    </li>
    <li
      aria-label="Slide 6 of 7"
      class="element"
      role="tab"
    >
      <div
        style="width: 20rem; height: 12rem; border-radius: 4px; background-color: rgb(238, 238, 238); display: flex; align-items: center; justify-content: center;"
      >
        Example 6
      </div>
    </li>
    <li
      aria-label="Slide 7 of 7"
      class="element"
      role="tab"
    >
      <div
        style="width: 20rem; height: 12rem; border-radius: 4px; background-color: rgb(238, 238, 238); display: flex; align-items: center; justify-content: center;"
      >
        Example 7
      </div>
    </li>
  </ul>
  <button
    class="wrapper wrapper-0-2-1 wrapper-d1-0-2-3 wrapper__clickable button"
    data-testid="right"
    tabindex="0"
  >
    <span
      class="jsx-2787895825 icon material-icons"
    >
      chevron_right
    </span>
  </button>
</div>
`;
